ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Version.hs

```diff
@@ -70,7 +70,7 @@ foldMapVersions :: (Ord vNum, Foldable f, HasCallStack)
                 -> f x
                 -> Versions vNum extra r
 foldMapVersions f fx = case toList fx of
-    [] -> error "foldMapVersions: precondition violated"
+    [] -> error "foldMapVersions: precondition violated" -- TODO
     xs -> foldl1 (<>) (map f xs)
```

Comment:
There is documentation of this precondition, but not of the rationale for it, as far as I can tell - and I note that the underlying structure can easily represent an empty collection of versions. This function, or the structure itself, may benefit from such documentation.

Reply:
I don't seem to be able to follow that link. Would you mind naming the file/files you're pointing me to?

Reply:
I would like to use `NonEmpty` list for this, but I've been waiting for the `Foldable1` type class, which we can use once we drop `ghc-8.10`. It's enough for now to create an issue for it or leave a `TODO` note.

Reply:
I've opened #4677 for this.